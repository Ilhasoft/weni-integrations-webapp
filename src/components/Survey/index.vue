<template>
  <div class="survey">
    <transition name="fade">
      <div v-if="isOpen" class="survey__content">
        <unnnic-input
          class="survey__content__input"
          v-model="answer"
          maxlength="255"
          size="sm"
          :label="$t('survey.content.question')"
        />

        <div class="survey__content__buttons">
          <unnnic-button
            class="survey__content__buttons__exit"
            type="tertiary"
            size="small"
            :text="$t('general.exit')"
            @click="closeSurvey"
          />

          <unnnic-button
            class="survey__content__buttons__send"
            type="secondary"
            size="small"
            :text="$t('general.send')"
            :loading="loadingSurveyAnswer"
            @click="sendSurvey"
          />
        </div>
      </div>
    </transition>

    <div class="survey__toggle" @click="toggleContent">
      <unnnic-icon-svg icon="question-circle-1" scheme="neutral-darkest" size="lg" />
    </div>
  </div>
</template>

<script>
  import { mapState, mapActions } from 'vuex';
  import { unnnicCallAlert } from '@weni/unnnic-system';

  export default {
    name: 'Survey',
    data() {
      return {
        isOpen: false,
        answer: '',
      };
    },
    computed: {
      ...mapState('survey', [
        'surveyStatus',
        'loadingSurveyAnswer',
        'errorSurveyAnswer',
        'surveyAnswerResult',
      ]),
    },
    mounted() {
      this.isOpen = this.surveyStatus === 'TO_ANSWER';
    },
    methods: {
      ...mapActions('survey', ['setSurveyStatus', 'submitSurveyAnswer']),
      closeSurvey() {
        this.isOpen = false;
        this.setSurveyStatus({ status: 'CLOSED' });
      },
      async sendSurvey() {
        const payload = {
          answer: this.answer,
        };

        await this.submitSurveyAnswer({ payload });

        if (this.errorSurveyAnswer) {
          this.callModal({
            type: 'Error',
            text: this.$t(`survey.error_submit_answer`),
          });
          return;
        }

        this.isOpen = false;
        this.setSurveyStatus({ status: 'ANSWERED' });
        this.answer = '';

        this.callModal({
          type: 'Success',
          text: this.$t(`survey.success_submit_answer`),
        });
      },
      toggleContent() {
        this.isOpen = !this.isOpen;

        if (this.surveyStatus !== 'ANSWERED') {
          this.setSurveyStatus({ status: this.isOpen ? 'TO_ANSWER' : 'CLOSED' });
        }
      },
      callModal({ text, type }) {
        unnnicCallAlert({
          props: {
            text: text,
            title: type === 'Success' ? this.$t('general.success') : this.$t('general.error'),
            icon: type === 'Success' ? 'check-circle-1-1' : 'alert-circle-1',
            scheme: type === 'Success' ? 'feedback-green' : 'feedback-red',
            position: 'bottom-right',
            closeText: this.$t('general.Close'),
          },
          seconds: 6,
        });
      },
    },
  };
</script>

<style lang="scss" scoped>
  @import './styles.scss';
</style>
