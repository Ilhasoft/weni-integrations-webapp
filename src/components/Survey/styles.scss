.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.3s;
}
.fade-enter,
.fade-leave-to {
  opacity: 0;
}

.survey {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-stack-xs;

  &__content {
    display: flex;
    max-width: 300px;
    flex-direction: column;
    gap: $unnnic-spacing-stack-sm;
    background-color: $unnnic-color-background-grass;
    padding: $unnnic-squish-sm;
    border-radius: $unnnic-border-radius-sm;
    box-shadow: $unnnic-shadow-level-near;

    &__buttons {
      display: flex;
      gap: $unnnic-spacing-inline-nano;
      justify-content: center;

      &__exit,
      &__send {
        width: 100px;
      }
    }
  }

  &__toggle {
    display: flex;
    background-color: $unnnic-color-background-grass;
    width: 50px;
    aspect-ratio: 1;
    margin-left: auto;
    border-radius: $unnnic-border-radius-pill;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    box-shadow: $unnnic-shadow-level-near;

    ::v-deep .unnnic-icon {
      width: $unnnic-icon-size-xl;
      height: $unnnic-icon-size-xl;
    }
  }
}