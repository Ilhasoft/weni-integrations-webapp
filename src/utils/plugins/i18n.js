import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
