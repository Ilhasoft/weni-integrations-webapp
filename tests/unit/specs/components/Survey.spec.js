import { unnnicCallAlert as mockUnnnicCallAlert } from '@weni/unnnic-system';

jest.mock('@weni/unnnic-system', () => ({
  ...jest.requireActual('@weni/unnnic-system'),
  unnnicCallAlert: jest.fn(),
}));

import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Survey from '@/components/Survey/index.vue';
import i18n from '@/utils/plugins/i18n';

const localVue = createLocalVue();
localVue.use(Vuex);

const mountComponent = async ({
  surveyStatus = 'TO_ANSWER',
  loadingSurveyAnswer = false,
  errorSurveyAnswer = null,
  surveyAnswerResult = null,
} = {}) => {
  const state = {
    surveyStatus,
    loadingSurveyAnswer,
    errorSurveyAnswer,
    surveyAnswerResult,
  };

  const actions = {
    setSurveyStatus: jest.fn(),
    submitSurveyAnswer: jest.fn(),
  };

  const store = new Vuex.Store({
    modules: {
      survey: {
        namespaced: true,
        state,
        actions,
      },
    },
  });

  const wrapper = mount(Survey, {
    localVue,
    store,
    i18n,
  });

  await wrapper.vm.$nextTick();

  return { wrapper, actions, state };
};

describe('components/Survey.vue', () => {
  it('should be rendered properly with content open', async () => {
    let { wrapper } = await mountComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('should be rendered properly with content closed', async () => {
    let { wrapper } = await mountComponent({ surveyStatus: 'ANSWERED' });
    expect(wrapper).toMatchSnapshot();
  });

  it('should invert isOpen to false on toggle click', async () => {
    const { wrapper } = await mountComponent();

    expect(wrapper.vm.isOpen).toEqual(true);

    let toggleButton = wrapper.find('.survey__toggle');
    await toggleButton.trigger('click');

    expect(wrapper.vm.isOpen).toEqual(false);

    let content = wrapper.find('.survey__content');
    expect(content.exists()).toBe(false);
  });

  it('should call setSurveyStatus to CLOSED if user never answered and content survey is open', async () => {
    const { wrapper, actions } = await mountComponent();

    expect(wrapper.vm.isOpen).toEqual(true);

    let toggleButton = wrapper.find('.survey__toggle');
    await toggleButton.trigger('click');

    expect(wrapper.vm.isOpen).toEqual(false);

    expect(actions.setSurveyStatus).toHaveBeenCalledTimes(1);
    expect(actions.setSurveyStatus).toHaveBeenCalledWith(expect.any(Object), { status: 'CLOSED' });

    let content = wrapper.find('.survey__content');
    expect(content.exists()).toBe(false);
  });

  it('should call setSurveyStatus to TO_ANSWER if user never answered and content survey is closed', async () => {
    const { wrapper, actions } = await mountComponent({ surveyStatus: 'CLOSED' });

    expect(wrapper.vm.isOpen).toEqual(false);

    let toggleButton = wrapper.find('.survey__toggle');
    await toggleButton.trigger('click');

    expect(wrapper.vm.isOpen).toEqual(true);

    expect(actions.setSurveyStatus).toHaveBeenCalledTimes(1);
    expect(actions.setSurveyStatus).toHaveBeenCalledWith(expect.any(Object), {
      status: 'TO_ANSWER',
    });

    let content = wrapper.find('.survey__content');
    expect(content.exists()).toBe(true);
  });

  it('should not call setSurveyStatus if user has already answered', async () => {
    const { wrapper, actions } = await mountComponent({ surveyStatus: 'ANSWERED' });

    expect(wrapper.vm.isOpen).toEqual(false);

    let toggleButton = wrapper.find('.survey__toggle');
    await toggleButton.trigger('click');

    expect(wrapper.vm.isOpen).toEqual(true);

    expect(actions.setSurveyStatus).not.toHaveBeenCalled();

    let content = wrapper.find('.survey__content');
    expect(content.exists()).toBe(true);
  });

  it('should set isOpen as closed on exit click', async () => {
    const { wrapper } = await mountComponent();

    expect(wrapper.vm.isOpen).toEqual(true);

    let closeButton = wrapper.find('.survey__content__buttons__exit');
    await closeButton.trigger('click');

    expect(wrapper.vm.isOpen).toEqual(false);

    const content = wrapper.find('.survey__content');
    expect(content.exists()).toBe(false);
  });

  it('should send answer and close content', async () => {
    const { wrapper, actions } = await mountComponent();
    const alertSpy = spyOn(wrapper.vm, 'callModal');

    expect(wrapper.vm.isOpen).toEqual(true);

    await wrapper.setData({ answer: 'answer content' });

    const sendButton = wrapper.find('.survey__content__buttons__send');
    await sendButton.trigger('click');
    await wrapper.vm.$nextTick();

    expect(actions.submitSurveyAnswer).toHaveBeenCalledWith(expect.any(Object), {
      payload: { answer: 'answer content' },
    });

    expect(alertSpy).toHaveBeenCalledTimes(1);
    expect(alertSpy).toHaveBeenCalledWith({ type: 'Success', text: 'Feedback successfully sent!' });

    expect(wrapper.vm.isOpen).toEqual(false);
    const content = wrapper.find('.survey__content');
    expect(content.exists()).toBe(false);
  });

  it('should try send answer and alert on error', async () => {
    const { wrapper, actions } = await mountComponent({ errorSurveyAnswer: true });
    const alertSpy = spyOn(wrapper.vm, 'callModal');

    expect(wrapper.vm.isOpen).toEqual(true);

    await wrapper.setData({ answer: 'answer content' });

    const sendButton = wrapper.find('.survey__content__buttons__send');
    await sendButton.trigger('click');
    await wrapper.vm.$nextTick();

    expect(actions.submitSurveyAnswer).toHaveBeenCalledWith(expect.any(Object), {
      payload: { answer: 'answer content' },
    });

    expect(alertSpy).toHaveBeenCalledTimes(1);
    expect(alertSpy).toHaveBeenCalledWith({
      type: 'Error',
      text: 'Failed to submit feedback, please try again later.',
    });

    expect(wrapper.vm.isOpen).toEqual(true);
    const content = wrapper.find('.survey__content');
    expect(content.exists()).toBe(true);
  });

  describe('callModal', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('should call unnnicCallAlert with Success state', async () => {
      const { wrapper } = await mountComponent();

      expect(mockUnnnicCallAlert).not.toHaveBeenCalled();

      wrapper.vm.callModal({ text: 'success', type: 'Success' });

      expect(mockUnnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'success',
          title: 'Success',
          icon: 'check-circle-1-1',
          scheme: 'feedback-green',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 6,
      });
    });

    it('should call unnnicCallAlert with Error state', async () => {
      const { wrapper } = await mountComponent();

      expect(mockUnnnicCallAlert).not.toHaveBeenCalled();

      wrapper.vm.callModal({ text: 'error', type: 'Error' });

      expect(mockUnnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'error',
          title: 'Error',
          icon: 'alert-circle-1',
          scheme: 'feedback-red',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 6,
      });
    });
  });
});
