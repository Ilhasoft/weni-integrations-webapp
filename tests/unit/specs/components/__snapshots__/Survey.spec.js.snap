// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Survey.vue should be rendered properly with content closed 1`] = `
<div class="survey">
  <transition-stub name="fade">
    <!---->
  </transition-stub>
  <div class="survey__toggle"><span data-v-59b073b4="" class="unnnic-icon unnnic-icon__size--lg unnnic-icon-scheme--neutral-darkest"><svg viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M20 23.7505C19.4825 23.7505 19.0625 23.3305 19.0625 22.813V21.5555C19.0612 20.3655 19.815 19.2992 20.9375 18.903C22.0588 18.5067 22.8125 17.4405 22.8125 16.2517C22.8125 15.5005 22.52 14.7942 21.9887 14.263C21.4587 13.7317 20.7525 13.4392 20 13.4392C19.2487 13.4392 18.5425 13.7317 18.0113 14.263C17.48 14.793 17.1875 15.4992 17.1875 16.2505C17.1875 16.7667 16.7675 17.188 16.25 17.188C15.7325 17.188 15.3125 16.7667 15.3125 16.2505C15.3125 14.998 15.8012 13.8217 16.6862 12.9367C17.5712 12.0517 18.7487 11.5642 20 11.5642C22.585 11.5642 24.6875 13.668 24.6875 16.2517C24.6875 18.2342 23.4312 20.0092 21.5625 20.6705C21.1887 20.803 20.9375 21.158 20.9375 21.5542V22.813C20.9375 23.3305 20.5175 23.7505 20 23.7505Z" fill="#3B414D" class="primary"></path><path d="M20 28.438C20.7767 28.438 21.4062 27.8084 21.4062 27.0317C21.4062 26.2551 20.7767 25.6255 20 25.6255C19.2233 25.6255 18.5938 26.2551 18.5938 27.0317C18.5938 27.8084 19.2233 28.438 20 28.438Z" fill="#3B414D" class="primary"></path><path d="M20 35.0005C11.7287 35.0005 5 28.2717 5 20.0005C5 11.7292 11.7287 5.00049 20 5.00049C28.2712 5.00049 35 11.7292 35 20.0005C35 28.2717 28.2712 35.0005 20 35.0005ZM20 6.87549C12.7625 6.87549 6.875 12.763 6.875 20.0005C6.875 27.238 12.7625 33.1255 20 33.1255C27.2375 33.1255 33.125 27.238 33.125 20.0005C33.125 12.763 27.2375 6.87549 20 6.87549Z" fill="#3B414D" class="primary"></path></svg></span></div>
</div>
`;

exports[`components/Survey.vue should be rendered properly with content open 1`] = `
<div class="survey">
  <transition-stub name="fade">
    <div class="survey__content">
      <div data-v-48c0c241="" class="survey__content__input unnnic-form sm" maxlength="255">
        <p data-v-48c0c241="" class="unnnic-form__label">Do you miss any integration? If so, tell us which one</p>
        <div data-v-400546b3="" data-v-48c0c241="" class="unnnic-form-input text-input size--sm" v-bind="[object Object]" mask=""><input data-v-97b44e8c="" data-v-400546b3="" type="text" size="0" v-bind="[object Object]" maxlength="255" placeholder="" class="input-itself input size-sm normal">
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <div class="survey__content__buttons"><button data-v-c21b2f22="" class="survey__content__buttons__exit unnnic-button unnnic-button--size-small unnnic-button--terciary unnnic-button-scheme--">
          <!---->
          <!---->
          <!----><span data-v-c21b2f22="" class="unnnic-button__label"> Exit </span>
          <!---->
        </button> <button data-v-c21b2f22="" class="survey__content__buttons__send unnnic-button unnnic-button--size-small unnnic-button--secondary unnnic-button-scheme--">
          <!---->
          <!---->
          <!----><span data-v-c21b2f22="" class="unnnic-button__label"> Send </span>
          <!---->
        </button></div>
    </div>
  </transition-stub>
  <div class="survey__toggle"><span data-v-59b073b4="" class="unnnic-icon unnnic-icon__size--lg unnnic-icon-scheme--neutral-darkest"><svg viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M20 23.7505C19.4825 23.7505 19.0625 23.3305 19.0625 22.813V21.5555C19.0612 20.3655 19.815 19.2992 20.9375 18.903C22.0588 18.5067 22.8125 17.4405 22.8125 16.2517C22.8125 15.5005 22.52 14.7942 21.9887 14.263C21.4587 13.7317 20.7525 13.4392 20 13.4392C19.2487 13.4392 18.5425 13.7317 18.0113 14.263C17.48 14.793 17.1875 15.4992 17.1875 16.2505C17.1875 16.7667 16.7675 17.188 16.25 17.188C15.7325 17.188 15.3125 16.7667 15.3125 16.2505C15.3125 14.998 15.8012 13.8217 16.6862 12.9367C17.5712 12.0517 18.7487 11.5642 20 11.5642C22.585 11.5642 24.6875 13.668 24.6875 16.2517C24.6875 18.2342 23.4312 20.0092 21.5625 20.6705C21.1887 20.803 20.9375 21.158 20.9375 21.5542V22.813C20.9375 23.3305 20.5175 23.7505 20 23.7505Z" fill="#3B414D" class="primary"></path><path d="M20 28.438C20.7767 28.438 21.4062 27.8084 21.4062 27.0317C21.4062 26.2551 20.7767 25.6255 20 25.6255C19.2233 25.6255 18.5938 26.2551 18.5938 27.0317C18.5938 27.8084 19.2233 28.438 20 28.438Z" fill="#3B414D" class="primary"></path><path d="M20 35.0005C11.7287 35.0005 5 28.2717 5 20.0005C5 11.7292 11.7287 5.00049 20 5.00049C28.2712 5.00049 35 11.7292 35 20.0005C35 28.2717 28.2712 35.0005 20 35.0005ZM20 6.87549C12.7625 6.87549 6.875 12.763 6.875 20.0005C6.875 27.238 12.7625 33.1255 20 33.1255C27.2375 33.1255 33.125 27.238 33.125 20.0005C33.125 12.763 27.2375 6.87549 20 6.87549Z" fill="#3B414D" class="primary"></path></svg></span></div>
</div>
`;
